New: `no-unused-modules` rule (#1142)
benmosher/eslint-plugin-import: docs/rules/no-unused-modules.md

# import/no-unused-modules

Reports:

>   - modules without any or

Reviewer: modules without any what?

>   - individual exports not being used within other modules

Reviewer: ```suggestion
  - individual exports not being statically `import`ed or `require`ed from other modules in the same project
```

Note: dynamic imports are currently not supported.

## Rule Details

### Options

This rule takes the following option:

> - `src`: an array with files/paths to be analyzed. It only for applies for unused exports. Defaults to `process.cwd()`, if not provided

Reviewer: ```suggestion
- `src`: an array with files/paths to be analyzed. It only applies to unused exports. Defaults to `process.cwd()`, if not provided
```

> - `ignoreExports`: an array with files/paths for which unused exports will not be reported (e.g module entry points) 

Reviewer: ```suggestion
- `ignoreExports`: an array with files/paths for which unused exports will not be reported (e.g module entry points in a published package)
```

- `missingExports`: if `true`, files without any exports are reported  

> - `unusedExports`: if `true`, exports without any usage within other modules are reported.

Reviewer: If both missingExports and unusedExports are missing or false, what gets reported?

If nothing, can we ensure it’s a schema error if neither of them are true?

Author: I'm lacking experience with the JSON schema. Do you have an idea or example on how to conditionally require one or the other field?

Author: Finally figured it out. We now get a schema error, if both are missing/false.

> - `unusedExports`: if `true`, exports without any usage within other modules are reported.

Reviewer: ```suggestion
- `unusedExports`: if `true`, exports without any static usage within other modules are reported.
```

export { foo, bar }
```
```js
export { foo as bar }
```

### Example for unused exports
given file-f:
```js
import { e } from 'file-a'
import { f } from 'file-b'
import * from  'file-c'
export * from 'file-d'
export { default, i0 } from 'file-e' // both will be reported

export const j = 99 // will be reported 
```
and file-e:
```js
export const i0 = 9 // will not be reported
export const i1 = 9 // will be reported
export default () => {} // will not be reported
```
and file-d:
```js
export const h = 8 // will not be reported
export default () => {} // will be reported, as export * only considers named exports and ignores default exports
```
and file-c:
```js
export const g = 7 // will not be reported
export default () => {} // will not be reported
```
and file-b:
```js
import two, { b, c, doAnything } from 'file-a'

export const f = 6 // will not be reported
```
and file-a:
```js
const b = 2
const c = 3
const d = 4

export const a = 1 // will be reported

export { b, c } // will not be reported

export { d as e } // will not be reported

export function doAnything() {
  // some code
}  // will not be reported

export default 5 // will not be reported
```

## When not to use

> If you don't mind having unused files or dead code within your codebase, you can disable this rule

Reviewer: All files should have a trailing newline

Author: Newline added with 2922910.

> If you don't mind having unused files or dead code within your codebase, you can disable this rule

Reviewer: so passive-aggressive 🤣

Author: Yeah...well...that's kind of a occupational disease I fear :scream:

What about ```If you have intentionally unused modules within your project, you can disable this rule```?

Reviewer: lol it's fine as-is, i just thought it was funny :-)

# import/no-unused-modules

Reports:
  - modules without any or
  - individual exports not being used within other modules

## Rule Details

### Options

This rule takes the following option:

> - `src`: an array with files/paths to be analyzed. It only for applies for unused exports  

Reviewer: i'm confused about this one - rule schemas don't generally specify their own paths, eslint itself does.

Author: For my knowledge, eslint does not share this information with the rules - but I need it for building up a list of existing exports and imports for this rule to work.

If there is a way to get this information from eslint, I would actually be glad to get rid of this option.

Reviewer: I'm confused why this wouldn't just use "the whole project" - ie, `process.cwd()` recursively.

Author: In my opinion, both solutions aren't optimal...

If we would use `process.cwd()`, the rule would always start its analysis within the root directory, even if eslint is restricted to a single directory within the root directory. In this case, one could of course use the `ignore`-option to restrict the files being linted...but then one could as well just maintain the `src`-option providing the one directory he intends to lint.

Author: Good idea. @ljharb: what do you think about this?

Reviewer: that seems more reasonable.

Author: I added the default src with aa7253baf383f370b3dccefbb11b17cd5db2d8dc. Readme will be updated later on.

- `ignore`: an array with files/paths to be ignored. It only for applies for unused exports  
- `missingExports`: if `true`, files without any exports are reported  

> - `unusedExports`: if `true`, exports without any usage are reported

Author: Thanks for the input. I will update the documentation.

> - `unusedExports`: if `true`, exports without any usage are reported

Reviewer: I'm confused, why would you ever want to allow an export that's only used in the same file and never imported?

> - `unusedExports`: if `true`, exports without any usage are reported

Reviewer 2: I was misunderstanding the purpose of the `ignore` option.  I had expected that `ignore: ['test/**']` would tell this rule to ignore imports performed by tests - thus report anything which is not imported from production code.  A common pattern in my applications is for internal sources to contain `export function defaultOpts() { return {...}; }`, and for that function to be used in exactly two places: the constructor function in the same source and the unit test.

To be honest I don't understand the purpose of the `ignore` option as it's currently implemented.  Why not just use .eslintrc `overrides` to turn this rule off for certain files?  Please correct me if I'm wrong but the `src` option doesn't tell this rule which files to lint, it tells it which files scan for import's, then each file being linted is checked to make sure something from `src` performed the import?  If so I think that `ignore` should counter-act `src`, meaning restrict the list of files scanned for import's - not restrict the list of files to lint.

> - `unusedExports`: if `true`, exports without any usage are reported

Reviewer: The option is for package entry points, which are often untested (because they just re-export something from the build dir).

I expect the spec option to cover files whose *exports* should be checked. Anything that imports it anywhere should count it as used.

> - `unusedExports`: if `true`, exports without any usage are reported

Author: As @ljharb said, the ```ignore``` option is mainly for package entry points. Every file specified in this option will not be linted by this rule. 
However, these files are also analyzed to handle their imports. This means, if an ignored file imports from anywhere else, let's say from ```file-a.js``` the corresponding exports from file  ```file-a.js```  will still be marked as used.

> - `unusedExports`: if `true`, exports without any usage are reported

Reviewer 2: I misunderstood the purpose of the `ignore` option, IMO it would be clearer if the docs mentioned package entry points.  Renaming it to `ignoreExports` or `entryPoints` could make it's purpose even clearer.

Would you accept a follow-up patch to implement an `ignoreImports` option?  This would allow me to identify when an API is no longer used even though it's still covered by unit tests.  I do test driven development so every export will be imported (directly or indirectly) by unit tests.

Thank you for your responses!

> - `unusedExports`: if `true`, exports without any usage are reported

Reviewer: I agree; both options are needed to be maximally useful.

> - `unusedExports`: if `true`, exports without any usage are reported

Reviewer 2: Nothing was ever changed in response to this comment thread.  I'm fine with pushing a follow-up PR to add support for ignoring importers (like tests for example), but I think naming the option `ignore` is confusing.

If you don't want to rename the option maybe some rewording of the docs?  Especially the docs for `ignore`, anyone trying to ignore tests will see the current docs and assume that it allows them to ignore importers (tests), rather than ignoring module entry-points.
```
- `src`: an array with files/paths with exports that must be used.  Defaults to `process.cwd()`, if not provided
- `ignore`: an array with files/paths where unused exports are ignored.
```

> - `unusedExports`: if `true`, exports without any usage are reported

Author: @coreyfarrell: sorry, that wasn't intentionally. I will rename the ```ignore```-option to ```ignoreImports``` and do some rewording in the docs.
